src/components/wizard_steps/MirrorSelect.vue

```diff
@@ -93,63 +118,70 @@ export default {
-    tools_mirrors: [],
-    pypi_mirrors: [],
-    defaultMirrors: {
-      idf: '',
-      tools: '',
-      pypi: ''
+    autoSelect: {
+      idf: true,
+      tools: true,
+      pypi: true
     }
   }),
   methods: {
-    get_available_idf_mirrors: async function () {
-      const idf_mirrors = await invoke("get_idf_mirror_list", {});
-      this.idf_mirrors = idf_mirrors.mirrors.map((mirror, index) => {
-        return {
-          value: mirror,
-          label: mirror,
-        }
-      });
-      this.selected_idf_mirror = idf_mirrors.selected;
-      this.loading_idfs = false;
-      return false;
+    buildMirrorList(type) {
+      const urls = type === 'idf' ? this.mirrorsStore.idf_urls : type === 'tools' ? this.mirrorsStore.tools_urls : this.mirrorsStore.pypi_urls;
+      const entries = type === 'idf' ? this.mirrorsStore.idf_entries : type === 'tools' ? this.mirrorsStore.tools_entries : this.mirrorsStore.pypi_entries;
+      if (Array.isArray(entries) && entries.length > 0) {
+        return entries.map(e => ({
+          value: e.url,
+          label: e.url,
+          ping: (e && Object.prototype.hasOwnProperty.call(e, 'latency')) ? (e.latency == null ? 0 : Number(e.latency)) : null,
+        }));
+      }
+      return (urls || []).map((url) => ({
+        value: url,
+        label: url,
+        ping: null,
+      }));
     },
-    get_available_tools_mirrors: async function () {
-      const tools_mirrors = await invoke("get_tools_mirror_list", {});
-      this.tools_mirrors = tools_mirrors.mirrors.map((mirror, index) => {
-        return {
-          value: mirror,
-          label: mirror,
-        }
-      });
-      this.selected_tools_mirror = tools_mirrors.selected;
-      this.loading_tools = false;
-      return false;
+    onSelectChange(type) {
+      // User has manually chosen a mirror for this type; stop auto-selecting
+      if (this.autoSelect[type]) {
+        this.autoSelect[type] = false;
+      }
     },
-    get_available_pypi_mirrors: async function () {
-      const pypi_mirrors = await invoke("get_pypi_mirror_list", {});
-      this.pypi_mirrors = pypi_mirrors.mirrors.map((mirror, index) => {
-        return {
-          value: mirror,
-          label: mirror,
-        }
-      });
-      this.selected_pypi_mirror = pypi_mirrors.selected;
-      this.loading_pypi = false;
-      return false;
+    getBestMirror(list) {
+      // best = smallest positive ping; ignore null (unknown) and 0 (timeout)
+      const candidates = list.filter(m => m.ping !== null && m.ping > 0);
+      if (candidates.length === 0) return null;
+      let best = candidates[0];
+      for (let i = 1; i < candidates.length; i++) {
+        if (candidates[i].ping < best.ping) best = candidates[i];
+      }
+      return best;
+    },
+    maybeAutoSelectBest(type) {
+      if (!this.autoSelect[type]) return;
+      const list = type === 'idf' ? this.idf_mirrors : type === 'tools' ? this.tools_mirrors : this.pypi_mirrors;
```

Comment:
this can be written more readably like: `const list   = this[`${type}_mirrors`];`